### DISCUSSION TITLE
TypeError: issubclass() arg 1 must be a class when inheriting BaseModel of a class with a complex type

### DISCUSSION THREAD

**Discussion by ValiaIsNotProgrammer** [fastapi/sqlmodel#949] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the SQLModel documentation, with the integrated search.
- [X] I already searched in Google "How to X in SQLModel" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to SQLModel but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to SQLModel but to [SQLAlchemy](https://github.com/sqlalchemy/sqlalchemy).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
class File(BaseModel):
    name: str
    url: str

class Producer(BaseModel):
    . . .
    files: Optional[List[File]]


class FileSQL(SQLModel, File, table=True):
    . . . 
    producer: Optional["ProducerSQL"] = Relationship(back_populates="files")


class ProducerSQL(SQLModel, Producer, table=True):
    . . .
    files: Optional[List[FileSQL]] = Relationship(back_populates="producer")
```

### Description

When attempting to define a relationship between two models (ProducerSQL and FileSQL) using SQLModel, an error is raised if the base model (Producer) contains a field of type Optional[List[File]]. The error does not occur if this field is removed.

#### Actual Behavior
```python
Traceback (most recent call last):
  File "<module>", line 83, in <module>
    class ProducerSQL(SQLModel, Producer, table=True):
  File "<virtualenv>/lib/python3.11/site-packages/sqlmodel/main.py", line 482, in __new__
    col = get_column_from_field(v)
          ^^^^^^^^^^^^^^^^^^^^^^^^
  File "<virtualenv>/lib/python3.11/site-packages/sqlmodel/main.py", line 626, in get_column_from_field
    sa_type = get_sqlalchemy_type(field)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<virtualenv>/lib/python3.11/site-packages/sqlmodel/main.py", line 575, in get_sqlalchemy_type
    if issubclass(type_, Enum):
       ^^^^^^^^^^^^^^^^^^^^^^^
TypeError: issubclass() arg 1 must be a class
```

#### Temporary solution
I obviously took out another intermediate model (fortunately, this one fit into my pipeline) and implemented the following method:
```python
@staticmethod
def prepare_to_db(item: ProducerLoader) -> ProducerSQL:
        files = [FileSQL(**file.dict()) for file in item.files]
        delattr(item, "files")
        sql_model = ProducerSQL(**item.dict())
        sql_model.files = files
        return sql_model
```
It's dirty, but working

### Operating System

Linux

### Operating System Details

Ubuntu 22.04.4 LTS

### SQLModel Version

0.0.18

### Python Version

3.11

### Additional Context

pydantic = 2.7.1
sqlalchemy = 2.0.30

**Comment by imneedle** [comment]
What I would suggest is to stay away from `BaseModel` entirely.

Remember that SQLModel is a wrapper around Pydantic and SQLAlechemy.

So behind the scenes, the `SQLModel` class inherits from `BaseModel`.

For the most part (at least in your example above), inheriting from `SQLModel` with `table=False` is the same thing as inheriting from `BaseModel`.

The fix I would suggest would be to always inherit from `SQLModel`, and specify `table=True` when you want it to be an SQL table.

This would also remove your multiple inheritance pattern like `ProducerSQL(SQLModel, Producer, table=True)`, and keep it as `ProducerSQL(Producer, table=True)` (assuming that `Producer` itself inherits from `SQLModel`)

```py
from typing import List, Optional

from sqlmodel import SQLModel

class File(SQLModel):  # table=False by default
    name: str
    url: str

class Producer(SQLModel):
    ...
    files: Optional[List[File]]


class FileSQL(File, table=True):  # `File` itself inherits from SQLModel
    ...
    producer: Optional["ProducerSQL"] = Relationship(back_populates="files")


class ProducerSQL(Producer, table=True):  # `Producer` itself inherits from SQLModel
    ...
    files: Optional[List[FileSQL]] = Relationship(back_populates="producer")
```

**Reply by ValiaIsNotProgrammer** [reply]
Indeed, in my case it costs nothing to completely replace BaseModel with SQLModel. However, this did not solve my problem when inheriting complex type
```python
class File(SQLModel):  # table=False by default
    name: str
    url: str

class Producer(SQLModel):
    files: Optional[List[File]]


class FileSQL(File, table=True):  # `File` itself inherits from SQLModel
    __tablename__ = "file"
    id: Optional[int] = Field(default=None, primary_key=True)
    producer_id: Optional[int] = Field(default=None, foreign_key="producer.id")
    producer: Optional["ProducerSQL"] = Relationship(back_populates="files")


class ProducerSQL(Producer, table=True):  # `Producer` itself inherits from SQLModel
    __tablename__ = "producer"
    id: Optional[int] = Field(default=None, primary_key=True)
    url: str = Field(unique=True)
    files: Optional[List['FileSQL']] = Relationship(back_populates="producer")
```
    
#### Traceback
```python
  File "/home/valiaisnotprogrammer/PycharmProjects/sql_model/inherit_complex_type", line 22, in <module>
    class ProducerSQL(Producer, table=True):  # `Producer` itself inherits from SQLModel
  File "/home/valiaisnotprogrammer/.cache/pypoetry/virtualenvs/flow-engineering-requests-3i-2j0Pi-py3.11/lib/python3.11/site-packages/sqlmodel/main.py", line 482, in __new__
    col = get_column_from_field(v)
          ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/valiaisnotprogrammer/.cache/pypoetry/virtualenvs/sql_model-3i-2j0Pi-py3.11/lib/python3.11/site-packages/sqlmodel/main.py", line 626, in get_column_from_field
    sa_type = get_sqlalchemy_type(field)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/valiaisnotprogrammer/.cache/pypoetry/virtualenvs/sql_model-3i-2j0Pi-py3.11/lib/python3.11/site-packages/sqlmodel/main.py", line 575, in get_sqlalchemy_type
    if issubclass(type_, Enum):
       ^^^^^^^^^^^^^^^^^^^^^^^
TypeError: issubclass() arg 1 must be a class
```

**Reply by imneedle** [reply, marked as answer]
Ah! That problem is linked to an error in your model definitions.

More specifically, it's because in your `Producer` class you try to set a type as `List[File]`. But `List` has no type in `sqlalchemy`.

In SQL databases, when you want to use lists, you typically have to use a `foreign_key`, or sometimes a `link_table`.

What I usually like to do is define `Create`, `Read` and `Update` models for my data. That way, it's easy later if you want to get `Read` data with relationship attributes for example.

I would think the best way to get what you want in your case is the following:

```py
from typing import Annotated, Optional

from sqlmodel import SQLModel, Field, Relationship


class FileCreate(SQLModel):
    name: str
    url: str
    # depends on how your API works, but typically, the file was created by a
    # producer, so you would have `producer_id` in the CreateModel` rather
    # than in the SQLModel... but that is up to you.


class FileSQL(FileCreate, table=True):
    __tablename__ = "file"
    id: Annotated[int | None, Field(primary_key=True)] = None
    producer_id: Annotated[int | None, Field(foreign_key="producer.id")] = None
    producer: Optional["ProducerSQL"] = Relationship(back_populates="files")


class FileRead(FileCreate):
    id: int
    producer_id: int


class ProducerCreate(SQLModel):
    url: Annotated[str, Field(unique=True)]


class ProducerSQL(ProducerCreate, table=True):
    __tablename__ = "producer"
    id: Annotated[int | None, Field(primary_key=True)] = None
    files: list["FileSQL"] = Relationship(back_populates="producer")


class ProducerRead(ProducerCreate):
    id: int


# Here, after you have defined all the Create, Read, Update and SQL models,
# you can define the read models with Relationships


class FileReadWithProducer(FileRead):
    producer: ProducerRead | None = None  # use the read model for Producer here


class ProducerReadWithFiles(ProducerRead):
    files: list[FileRead] = []  # use the read model for File here
```

And in your case, when you want to get the files for a producer, you make your code return the `ProducerReadWithFiles` Model.
Or you can use `FileReadWithProducer` if you want to get the producer of a file.

You can [find the documentation in SQLModel for this sort of problem](https://sqlmodel.tiangolo.com/tutorial/fastapi/relationships/).

**Comment by joachimhuet** [comment]
Hey! 

In fact the problem is not with the List but Rather with your Producer class that has a File type. 

By inheriting Producer the typing of `files` uses the File type that is not castable by sqlalchemy. 
So you should just avoid to declare relationship fields in base models as well as in table models:
```python
class File(SQLModel):  # table=False by default
    name: str
    url: str

# That would be the base for your producer object
class ProducerBase(SQLModel): ...

# the files declaration here does not conflict with the declaration in SQL model
class Producer(ProducerBase):
    files: Optional[List[File]]


class FileSQL(File, table=True):  # `File` itself inherits from SQLModel
    __tablename__ = "file"
    id: Optional[int] = Field(default=None, primary_key=True)
    producer_id: Optional[int] = Field(default=None, foreign_key="producer.id")
    producer: Optional["ProducerSQL"] = Relationship(back_populates="files")


class ProducerSQL(ProducerBase, table=True):  # `Producer` itself inherits from SQLModel
    __tablename__ = "producer"
    id: Optional[int] = Field(default=None, primary_key=True)
    url: str = Field(unique=True)
    files: Optional[List['FileSQL']] = Relationship(back_populates="producer")
```
